internal/contour/snapshot.go

```diff
@@ -40,7 +68,70 @@ func NewSnapshotHandler(c cache.SnapshotCache, resources []ResourceCache, logger
 	}
 }
 
-// OnChange is called when the DAG is rebuilt
-// and a new snapshot is needed.
+// OnChange is called when the DAG is rebuilt and a new snapshot is needed.
 func (s *SnapshotHandler) OnChange(root *dag.DAG) {
+
+	// Generate new snapshot version.
+	snapshotVersion := s.newSnapshotVersion()
+
+	// Create an snapshot with all xDS resources.
+	snapshot := cache.Snapshot{}
+
+	// Loop through all the xDS resources and pass them to the snapshot
+	for _, r := range s.resources {
+		snapshot.Resources[responseTypeOf(r.TypeURL())] = cache.NewResources(snapshotVersion, asResources(r.Contents()))
```

Comment:
How do you know it changed? At this point all we know is that the dag was rebuilt, so need to assume all the resources are different. We'd need to add a version to each resource type and then increment that if the content cache was updated. 

I think long term we should look at testing out ADS and reducing the gRCP load on Contour. Using that means we need a single version for the snapshot.

Reply:
> How do you know it changed? At this point all we know is that the dag was rebuilt, so need to assume all the resources are different. 

I was just thinking about something simple like comparing the contents to the last contents you have of that version.

You don't need a per-resource version, you can just update the resource to the latest global version (the version is opaque so it's OK to skip).

Reply:
But we'd need to keep a cache of the old version's contents to compare against the new or like I said keep a per-resource version and just compare against that. So if CDS last was version 10, and it's still version 10, then no changes have happened. 

Are you asking to only update the bits that change? If we want to move to ADS then we need to send all the caches together so that work is not required (assuming I understand you ask).

Reply:
> But we'd need to keep a cache of the old version's contents to compare against the new or like I said keep a per-resource version and just compare against that. So if CDS last was version 10, and it's still version 10, then no 
changes have happened.

You have the old contents in the snapshot already.

> Are you asking to only update the bits that change?

Yep, exactly.  Then we don't ask Envoy to do any work unless there are actual config changes. WE don't have to do it in this PR; maybe we can leave a comment and plan to experiment in a future issue.

>  If we want to move to ADS then we need to send all the caches together so that work is not required (assuming I understand you ask).

I don't think I'm following this. ADS is just using the same GRPC endpoint  for all the Envoy resource types. You still get a DiscoveryRequest for each resource type and AFAIK can version the types independently.